git sync

Felipe-gsilva opened the pull request:
fixes #257 , #371, #372 ,  #366 , #389 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced GitHub OAuth sign-in for a smoother authentication experience.
	- Expanded organization management in the dashboard, enabling viewing, editing, and inviting team members.
	- Added functionality for managing GitHub repository synchronization and organization details.
	- Implemented new components for user and organization management within the dashboard.
	- Enhanced mock management capabilities with new functions for retrieving and updating mocks based on organization details.

- **UI & Style Improvements**
	- Refreshed visual components with a modern color palette and updated button designs for enhanced usability.

- **Improvements**
	- Streamlined error messaging and data synchronization to deliver a more reliable, seamless user experience.
	- Enhanced logging and error handling across various components for better traceability and debugging.
	- Improved handling of user data, including additional fields for GitHub integration and organization management.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Felipe-gsilva committed:
feat: bumping npm and jar mockingbird into moclojer-app

deps.edn has a local reference for the .jar file for mockingbird, so
you'll need to build the jar and reference your own local path.

Felipe-gsilva committed:
feat: testing distribution of mockingbird

Felipe-gsilva committed:
feat: login pages with new mockingbird button

Felipe-gsilva committed:
feat: loading prop and fixes on example page

Felipe-gsilva committed:
fix: formatting and wip mockingbird css target

Felipe-gsilva committed:
feat: button and input on auth.views

Felipe-gsilva committed:
feat: adds buttons, images, pfps and messages from mockingbird

Felipe-gsilva committed:
fix: login image left roundness

Felipe-gsilva committed:
feat: last mockingbird components added

Felipe-gsilva committed:
feat: github signin

Felipe-gsilva committed:
fix: oauth instead of github

Felipe-gsilva committed:
feat: adds refx event for oauth

Felipe-gsilva committed:
fix: error and success status

Felipe-gsilva committed:
fix: moclojer icon on tab

Felipe-gsilva committed:
Merge branch 'feat/adding-mockingbird' into fix/minor-ui-issues

Felipe-gsilva committed:
refactor: semantic errors

Felipe-gsilva committed:
fix: removes mockingbird from npm

Felipe-gsilva committed:
fix: gh login pos

Felipe-gsilva committed:
fix: remomves some sensible keys from config.edn

Felipe-gsilva committed:
feat: initial sketch of webhook handler for syncing

Felipe-gsilva committed:
fix: correctly handling webhook

Felipe-gsilva committed:
feat: payload correct parsing

Felipe-gsilva committed:
feat (wip): auth with install id

Felipe-gsilva committed:
feat: pulling files

Felipe-gsilva committed:
feat: handling multiple files, mock filtering, and base64 encoding

Felipe-gsilva committed:
feat: retrieving ord id

Felipe-gsilva committed:
feat: updates mock, org and user to handle install id and repo link

this commit adds or modifies:
- schemas for customers, orgs and mocks
- correct config.edn context for git-sync
- installation process almost done

Felipe-gsilva committed:
feat(wip): pushing files via api and refactor

Felipe-gsilva committed:
feat: creating commits, trees and blobs for multiple mock pushing

Felipe-gsilva committed:
fix: updates mock sha on github

Felipe-gsilva committed:
chore: correct adapters and update on user and org

Felipe-gsilva committed:
fix: reverts public api keys removal

Felipe-gsilva committed:
refactor: git sync config

Felipe-gsilva committed:
Merge remote-tracking branch 'refs/remotes/origin/feat/git-sync' into feat/git-sync

Felipe-gsilva committed:
feat: settings section

Felipe-gsilva committed:
feat(wip): edit username. refactor: sync

Felipe-gsilva committed:
fix: git-install-id within wire and db context

Felipe-gsilva committed:
fix: migrations for git-install-id

Felipe-gsilva commented on src/back/api/logic/orgs.clj:
```diff
@@ -33,3 +33,6 @@
 (defn ->db-delete [{:keys [id]}]
   {:org/id (parse-uuid (str id))})
+
+(defn enable-sync [org install-id]
+  (assoc org {:org/git-install-id install-id}))
```
@coderabbitai this was already updated

Felipe-gsilva committed:
general fixes

Felipe-gsilva committed:
fix: events and local supabase

Felipe-gsilva committed:
fix: most of tests, url callback for git auth and migrations

Felipe-gsilva committed:
fix: edit-user flow

Felipe-gsilva committed:
fix: auth saving git_username

Felipe-gsilva commented on src/back/api/ports/http_in.clj:
```diff
@@ -200,3 +213,65 @@
+                                   updated-files))
+            pull-response (when mocks
+                            (controllers.sync/pull! installation-id git-slug repo-name mocks components ctx))
+            content (:content pull-response)
+            org (controllers.orgs/get-by-git-org-name git-slug components ctx)
+            org? (not (nil? (:orgname org)))
+            user (when-not org? (controllers.user/get-by-git-username git-slug components ctx))]
+        (if-not (nil? (:uuid (if org? org user)))
+          (do
+            (logs/log :info "mocks"
+                      :ctx (assoc ctx :mocks mocks))
+            (doseq [[mock-content sha] mocks]
+              (let [mock-id (controllers.mocks/get-mocks {:uuid (if org? org user)
+                                                          :username git-slug}
+                                                         components ctx)
+                    decoded-mock (utils/decode mock-content)]
+                (controllers.mocks/update-mock! mock-id decoded-mock sha components ctx)))
+            {:status 200
+             :body {:content content
+                    :message "Files updated from source"}})
+          {:status 500
+           :body {:message "no org or user found"}}))
+      (= event-type "installation")
+      (let [git-slug (get-in body [:installation :account :login])
+            org (controllers.orgs/get-by-git-org-name git-slug components ctx)
+            org? (not (nil? org))
+            user (when-not org? (controllers.user/get-by-git-username git-slug components ctx))]
+        (logs/log :info (str git-slug)
+                  :ctx (assoc ctx
+                              :id (if org? org user)))
+        (if-not (nil? (:uuid (if org? org user)))
+          {:status 200
+           :body {:message "Enabled Git Sync"
+                  :content (if org?
+                             (controllers.orgs/enable-sync installation-id org components ctx)
+                             (controllers.user/enable-sync installation-id user components ctx))}}
+          {:status 404
+           :body {:message "no org or user found"}}))
+      :else {:status 400
+             :body {:message "Unhandled event type"}})))
```
I do verify

Felipe-gsilva commented on src/front/app/auth/effects.cljs:
```diff
@@ -5,22 +5,32 @@
-                      (dispatch (conj on-success {:body (assoc
-                                                         (utils/js->cljs-key resp)
-                                                         :ok true)})))))))
+      (p/then #(handle-auth-response % on-success on-failure))
+      (p/catch (fn [resp]
+                 (js/console.log :resp resp)
+                 (dispatch (conj on-failure {:body (utils/js->cljs-key resp)}))))))
```
we need different responses, this suggestion makes no sense at all

Felipe-gsilva committed:
fix: user update and creation with oauth

Felipe-gsilva committed:
fix: installation and (wip) mock sync

Felipe-gsilva committed:
feat: creating mocks by webhook

Felipe-gsilva committed:
fix: correct content on mock creation

Felipe-gsilva committed:
feat(wip): handling multiple commits in one push

Felipe-gsilva committed:
feat: updating mocks

Felipe-gsilva committed:
feat (wip): pushing mock via frontend

Felipe-gsilva committed:
fix: default behavior on :put mock

Felipe-gsilva committed:
fix: correct req for pushing

Felipe-gsilva committed:
feat (wip): adding a repo into a mock

Felipe-gsilva committed:
feat: checking if user has git sync enabled

Felipe-gsilva committed:
feat: push in frontend and (wip) polling file from source for better UX

Felipe-gsilva committed:
feat: polling and reload file from server

Felipe-gsilva committed:
fix: org installation

Felipe-gsilva commented:
@avelino @J0sueTM  for you guys to know, the user can now use git sync, but not in prod, yet, because we will need to add a few things first (I mentioned it in #379 #378 #381 specially), but it is working within dev context (hit me out in discord so I can explain how to set things up, since we need some secrets for handling git ops

however, orgs are yet no perfectly implemented in UI ( as it should #377), for that I'm sorry, because things got a little messy with the scale of this feature here :pray: 

if anything seems off, just tag me here and I will explain it better for you all

Felipe-gsilva commented:
@coderabbitai give us a brief summary about what has been changed in this PR

Felipe-gsilva commented:
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> @Felipe-gsilva Here's a summary of the key changes in this PR:
> 
>     1. **Git Sync Integration**:
>        
>        * Added GitHub integration capabilities allowing mocks to be synced with git repositories
>        * Implemented GitHub App authentication and repository operations
>        * Added new database columns to support git sync (repo URLs, SHA hashes, install IDs)
> 
>     2. **User & Organization Updates**:
>        
>        * Enhanced user and organization models with GitHub-related fields
>        * Added support for GitHub usernames and organization names
>        * Implemented installation ID tracking for GitHub App
> 
>     3. **UI Enhancements**:
>        
>        * Added new settings modal for git sync configuration
>        * Implemented repository selection interface
>        * Updated mock editor with git sync controls
> 
>     4. **Dependencies & Configuration**:
>        
>        * Added GitHub-related dependencies (clj-github, data.json)
>        * Updated puppeteer to version 23.4.0
>        * Added git sync configuration section in config.edn
> 
> 
> Note: As mentioned, while the feature is functional in development, it requires additional work before production deployment, including proper organization UI implementation (#377) and other related issues (#379, #378, #381).

@coderabbitai you can be more specific over this answer, I said brief but I meant, not that brief

Felipe-gsilva committed:
fix (wip): overhaul on UX for git sync

Felipe-gsilva committed:
fix: ui flow and issues

Felipe-gsilva committed:
fix: config structure

Felipe-gsilva committed:
first commit: sketching org creation and listing

Felipe-gsilva committed:
feat: orgs indivdual views and events

Felipe-gsilva committed:
feat: org deletion and orgs-view

Felipe-gsilva committed:
fix: icons color and new svg

Felipe-gsilva committed:
fix: retriving orgs from user

avelino commented on src/back/api/ports/http_in.clj:
```diff
@@ -57,20 +64,29 @@
-                                             components
-                                             ctx)]
+  [{:keys [parameters session-data components ctx]}]
+  (let [user-id (:user-id session-data)
+        old-mock (:body parameters)
+        content (:content old-mock)
+        id (:id old-mock)
+        sha (:sha old-mock)
+        git-path (:git-repo old-mock)
+        git-repo (when git-path (last (str/split git-path #"/")))
+        new-mock (controllers.mocks/update-mock! (parse-uuid (str id))
+                                                 (-> {}
+                                                     (utils/assoc-if :content content)
+                                                     (cond->
+                                                      (not= sha "") (utils/assoc-if :sha sha)
+                                                      (not= git-repo "") (utils/assoc-if :git-repo git-path)))
+                                                 components
+                                                 ctx)]
     {:status 200
-     :body {:mock mock}}))
+     :body {:mock new-mock}}))
```
@Felipe-gsilva acredito que isso faça sentido, não validar só `""`

Felipe-gsilva replied:
@avelino realmente, eu nao coloquei a funcao q eu criei pra validar o sha ai :melting_face:

Felipe-gsilva committed:
fix: correct check for sha and github links

Felipe-gsilva committed:
fix: mock page, orgs specific mocks and general events issues

Felipe-gsilva committed:
fix: (wip)update orgname

Felipe-gsilva committed:
feat: get org specific mocks

Felipe-gsilva committed:
fix: org id on mock creation and mock display

Felipe-gsilva committed:
fix: user and org mock creation

Felipe-gsilva committed:
fix: add org user modal. feat: add existing user into org

Felipe-gsilva committed:
fix: correct request for adding users

Felipe-gsilva committed:
fix: overall issues

Felipe-gsilva committed:
feat: inviting new user by email

currently, this does not correctly redirects the user into the config
url, neither inserts a user into a org.

Felipe-gsilva committed:
fix: invite sending org-id

Felipe-gsilva committed:
fix: adds new user into org

it adds the user, but yet breaks the login flow

Felipe-gsilva committed:
fix: adding user into org by email

Felipe-gsilva committed:
fix: test and username update

Felipe-gsilva committed:
feat(wip): git-orgname update

Felipe-gsilva committed:
wip

Felipe-gsilva committed:
feat: (wip) git docs. fix: mock sync verify flow

Felipe-gsilva committed:
updated docs

Felipe-gsilva committed:
refactor: sync install and config flow

Felipe-gsilva committed:
Merge branch 'main' into feat/git-sync

Felipe-gsilva commented:
@coderabbitai clear all of your other reviews and redo it, considering the main aspects of the code. Do not give me too much assertive review, only the ones you judge necessary for the integrity of this PR

Felipe-gsilva commented on resources/back/config.edn:
```diff
@@ -6,6 +6,10 @@
  :supabase-api-key #or [#env SUPABASE_API_KEY "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6InRndmRmeHVyZ3NkZHhvdXhtdWdzIiwicm9sZSI6ImFub24iLCJpYXQiOjE2ODE1MDM0NDMsImV4cCI6MTk5NzA3OTQ0M30.7pq4MM_ZldiWvOk_cnQuxlvUF8eFcxlPDB7jMTNMYb0"]
```
@coderabbitai this key is a public

Felipe-gsilva committed:
docs: message on git sync

Felipe-gsilva committed:
fix: mobile message
